lib/src/log_handler.dart

```diff
@@ -0,0 +1,225 @@
+import 'dart:convert';
+
+import 'package:file/file.dart';
+import 'package:path/path.dart' as p;
+
+import 'constants.dart';
+import 'initializer.dart';
+
+/// Data class that will be returned when analyzing the
+/// persisted log file on the client's machine
+class LogFileStats {
+  /// The oldest timestamp in the log file
+  final DateTime startDateTime;
+
+  /// The latest timestamp in the log file
+  final DateTime endDateTime;
+
+  /// The number of unique session ids found in the log file
+  final int sessionCount;
+
+  /// The number of unique branches found in the log file
+  final int branchCount;
+
+  /// The number of unique tools found in the log file
+  final int toolCount;
+
+  /// Contains the data from the [LogHandler.logFileStats] method
+  const LogFileStats({
+    required this.startDateTime,
+    required this.endDateTime,
+    required this.sessionCount,
+    required this.branchCount,
+    required this.toolCount,
+  });
+
+  @override
+  String toString() => jsonEncode(<String, dynamic>{
+        'startDateTime': startDateTime.toString(),
+        'endDateTime': endDateTime.toString(),
+        'sessionCount': sessionCount,
+        'branchCount': branchCount,
+        'toolCount': toolCount,
+      });
+}
+
+/// This class is responsible for writing to a log
+/// file that has been initialized by the [Initializer]
+///
+/// It will be treated as an append only log and will be limited
+/// to have has many data records as specified by [kLogFileLength]
+class LogHandler {
+  final FileSystem fs;
+  final Directory homeDirectory;
+  final File logFile;
+
+  /// A log handler constructor that will delegate saving
+  /// logs and retrieving stats from the persisted log
+  LogHandler({
+    required this.fs,
+    required this.homeDirectory,
+  }) : logFile = fs.file(p.join(
+          homeDirectory.path,
+          kDartToolDirectoryName,
+          kLogFileName,
+        ));
+
+  /// Get stats from the persisted log file
+  LogFileStats? logFileStats() {
+    // Parse each line of the log file through [LogItem],
+    // some returned records may be null if malformed, they will be
+    // removed later through `whereType<LogItem>`
+    final List<LogItem> records = logFile
+        .readAsLinesSync()
+        .map((String e) => LogItem.fromRecord(jsonDecode(e)))
+        .whereType<LogItem>()
+        .toList();
+
+    if (records.isEmpty) return null;
+
+    // Get the start and end dates for the log file
+    final DateTime startDateTime = records.first.localTime;
+    final DateTime endDateTime = records.last.localTime;
+
```

Comment:
part 2 @christopherfujino  This is where it gets called, leaving us with a `List<LogItem>` that has no missing values